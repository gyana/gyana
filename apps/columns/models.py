from django.conf import settings
from django.core.validators import RegexValidator
from django.db import models

from apps.base.core.aggregations import AggregationFunctions
from apps.base.models import SaveParentModel
from apps.columns.currency_symbols import CurrencySymbols
from apps.nodes.models import Node

from .bigquery import (
    CommonOperations,
    DateOperations,
    DatePeriod,
    DatetimeOperations,
    NumericOperations,
    StringOperations,
    TimeOperations,
)

bigquery_column_regex = RegexValidator(
    r"^[a-zA-Z_][0-9a-zA-Z_]*$", "Only numbers, letters and underscores allowed."
)


class AbstractOperationColumn(SaveParentModel):
    class Meta:
        abstract = True

    column = models.CharField(max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH)

    string_function = models.CharField(
        max_length=20,
        choices=(
            (key, value.label)
            for key, value in {**CommonOperations, **StringOperations}.items()
        ),
        null=True,
        verbose_name="Operation",
    )
    integer_function = models.CharField(
        max_length=20,
        choices=(
            (key, value.label)
            for key, value in {**CommonOperations, **NumericOperations}.items()
        ),
        null=True,
        verbose_name="Operation",
    )
    date_function = models.CharField(
        max_length=20,
        choices=(
            (key, value.label)
            for key, value in {**CommonOperations, **DateOperations}.items()
        ),
        null=True,
        verbose_name="Operation",
    )
    time_function = models.CharField(
        max_length=20,
        choices=(
            (key, value.label)
            for key, value in {**CommonOperations, **TimeOperations}.items()
        ),
        null=True,
        verbose_name="Operation",
    )
    datetime_function = models.CharField(
        max_length=20,
        choices=(
            (key, value.label)
            for key, value in {
                **CommonOperations,
                **TimeOperations,
                **DateOperations,
                **DatetimeOperations,
            }.items()
        ),
        null=True,
        verbose_name="Operation",
    )

    boolean_function = models.CharField(
        max_length=20,
        choices=((key, value.label) for key, value in {**CommonOperations}.items()),
        null=True,
        verbose_name="Operation",
    )

    integer_value = models.BigIntegerField(null=True, verbose_name="Value")
    float_value = models.FloatField(null=True, verbose_name="Value")
    string_value = models.TextField(null=True, verbose_name="Value")

    @property
    def function(self):
        return (
            self.string_function
            or self.integer_function
            or self.date_function
            or self.time_function
            or self.datetime_function
        )


class ColumnSettings(models.Model):
    class Meta:
        abstract = True

    name = models.CharField(
        max_length=64,
        null=True,
        blank=True,
        help_text="Set column header name",
    )
    rounding = models.IntegerField(default=2, help_text="Decimal point to round to")
    currency = models.CharField(
        max_length=32,
        choices=CurrencySymbols.choices,
        blank=True,
        null=True,
        help_text="Select a currency",
    )
    is_percentage = models.BooleanField(
        default=False, blank=True, verbose_name="Show in %"
    )


class Column(ColumnSettings, SaveParentModel):
    column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH, help_text="Select columns"
    )
    node = models.ForeignKey(
        Node, null=True, on_delete=models.CASCADE, related_name="columns"
    )
    widget = models.ForeignKey(
        "widgets.Widget", on_delete=models.CASCADE, related_name="columns", null=True
    )
    part = models.CharField(
        max_length=16,
        choices=DatePeriod.choices,
        null=True,
        blank=True,
        help_text="Select the desired date part",
    )


class ConvertColumn(SaveParentModel):
    class TargetTypes(models.TextChoices):
        TEXT = "text", "Text"
        INT = "int", "Integer"
        FLOAT = "float", "Decimal"
        BOOL = "bool", "Bool"
        DATE = "date", "Date"
        TIME = "time", "Time"
        DATETIME = "timestamp", "Datetime"

    column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH, help_text="Select column"
    )
    target_type = models.CharField(
        max_length=16,
        choices=TargetTypes.choices,
        help_text="Select type to convert to",
    )
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="convert_columns"
    )


class SecondaryColumn(SaveParentModel):
    column = models.CharField(max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH)
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="secondary_columns"
    )


class AggregationColumn(ColumnSettings, SaveParentModel):
    class Meta:
        ordering = ("created",)

    column = models.CharField(max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH)
    function = models.CharField(max_length=20, choices=AggregationFunctions.choices)
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="aggregations", null=True
    )
    widget = models.ForeignKey(
        "widgets.Widget",
        on_delete=models.CASCADE,
        related_name="aggregations",
        null=True,
    )


class SortColumn(SaveParentModel):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="sort_columns"
    )
    ascending = models.BooleanField(default=True, help_text="Ascending Sort")
    column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
    )


class EditColumn(AbstractOperationColumn):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="edit_columns"
    )


class AddColumn(AbstractOperationColumn):
    node = models.ForeignKey(Node, on_delete=models.CASCADE, related_name="add_columns")
    label = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        validators=[bigquery_column_regex],
        verbose_name="New column name",
    )


class RenameColumn(SaveParentModel):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="rename_columns"
    )
    column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH, verbose_name="Old column name"
    )
    new_name = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        validators=[bigquery_column_regex],
        verbose_name="New column name",
    )


class FormulaColumn(SaveParentModel):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="formula_columns"
    )
    formula = models.TextField(null=True, help_text="Type formula")
    label = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        validators=[bigquery_column_regex],
        help_text="New column name",
    )


class WindowColumn(SaveParentModel):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="window_columns"
    )

    column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        help_text="Choose an aggregation column",
    )
    function = models.CharField(
        max_length=20,
        choices=AggregationFunctions.choices,
        help_text="Select an aggregation function",
    )
    group_by = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        null=True,
        blank=True,
        help_text="Group over this column",
    )
    order_by = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        null=True,
        blank=True,
        help_text="Order by this column",
    )
    ascending = models.BooleanField(
        default=True, help_text="Select to sort ascendingly"
    )
    label = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        validators=[bigquery_column_regex],
        help_text="Select a new column name",
    )


class JoinColumn(SaveParentModel):
    node = models.ForeignKey(
        Node, on_delete=models.CASCADE, related_name="join_columns"
    )
    how = models.CharField(
        max_length=12,
        choices=[
            ("inner", "Inner"),
            ("outer", "Outer"),
            ("left", "Left"),
            ("right", "Right"),
        ],
        default="inner",
        help_text="Select the join method, more information in the docs",
    )

    left_index = models.IntegerField(default=0)
    left_column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        help_text="Choose the join column from Input {}",
    )

    right_column = models.CharField(
        max_length=settings.BIGQUERY_COLUMN_NAME_LENGTH,
        help_text="Choose the right join column from Input {}",
    )
