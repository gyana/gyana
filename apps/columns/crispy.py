from crispy_forms.layout import TEMPLATE_PACK, TemplateNameMixin
from django.template.loader import render_to_string


class ColumnFormatting(TemplateNameMixin):
    template = "%s/layout/column_formatting.html"

    def __init__(self, name, fields):
        self.name = name
        self.fields = fields

    def render(self, form, context, template_pack=TEMPLATE_PACK, **kwargs):
        template = self.get_template_name(template_pack)
        context.update(
            {
                "name": form[self.name],
                "columns": [
                    (form[x], form[y])
                    for x, y in zip(self.fields[::2], self.fields[1::2])
                ],
            }
        )
        return render_to_string(template, context.flatten())
