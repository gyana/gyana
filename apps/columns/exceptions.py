class FunctionNotFound(Exception):
    def __init__(self, function) -> None:
        message = f"Function '{function}' does not exist"
        self.function = function
        super().__init__(message)


class ColumnNotFound(Exception):
    def __init__(self, column, columns) -> None:
        message = f"Column '{column}' does not exist"
        self.column = column
        self.columns = columns
        super().__init__(message)


class ArgumentError(Exception):
    def __init__(self, function, args):
        super().__init__()
        self.function = function
        self.args = args


class ColumnAttributeError(Exception):
    def __init__(self, column, function):
        super().__init__()
        self.column = column
        self.function = function


class ParseError(Exception):
    def __init__(self, formula, columns):
        super().__init__()
        self.formula = formula
        self.columns = columns
