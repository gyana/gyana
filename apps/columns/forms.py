from django import forms
from ibis.expr.datatypes import Array, Date, Floating, Integer, Struct, Timestamp

from apps.base.core.aggregations import AGGREGATION_TYPE_MAP
from apps.base.core.utils import create_column_choices
from apps.base.forms import BaseLiveSchemaForm, LiveAlpineModelForm, SchemaFormMixin
from apps.base.widgets import Datalist, SelectWithDisable
from apps.columns.models import (
    AddColumn,
    AggregationColumn,
    Column,
    ConvertColumn,
    EditColumn,
    FormulaColumn,
    JoinColumn,
    RenameColumn,
    SortColumn,
    WindowColumn,
)
from apps.widgets.models import Widget

from .bigquery import AllOperations, DatePeriod
from .widgets import CodeMirror

IBIS_TO_FUNCTION = {
    "String": "string_function",
    "Int8": "integer_function",
    "Int16": "integer_function",
    "Int32": "integer_function",
    "Int64": "integer_function",
    "Float64": "integer_function",
    "Timestamp": "datetime_function",
    "Date": "date_function",
    "Time": "time_function",
    "Boolean": "boolean_function",
}

FUNCTION_TO_IBIS = {
    "string_function": ["String"],
    "integer_function": ["Int8", "Int16", "Int32", "Int64", "Float64"],
    "datetime_function": ["Timestamp"],
    "date_function": ["Date"],
    "time_function": ["Time"],
    "boolean_function": ["Boolean"],
}


def disable_struct_and_array_columns(fields, column_field, schema):
    fields["column"] = forms.ChoiceField(
        choices=column_field.choices,
        help_text=column_field.help_text,
        widget=SelectWithDisable(
            disabled={
                name: "Currently, you cannot use this column type here."
                for name, type_ in schema.items()
                if isinstance(type_, (Struct, Array))
            }
        ),
    )


class ColumnForm(SchemaFormMixin, LiveAlpineModelForm):
    class Meta:
        fields = ("column", "part")
        model = Column

        show = {
            "part": "['Date', 'Timestamp'].includes(schema[column])",
        }

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        disable_struct_and_array_columns(
            self.fields, self.fields["column"], self.schema
        )

        date_periods = {
            "Date": [DatePeriod.DATE.value],
            "Timestamp": [x.value for x in DatePeriod],
        }
        self.fields["column"].widget.attrs.update(
            {"x-effect": f"$data.part_choices = {date_periods}[schema[column]]"}
        )


class ColumnFormWithFormatting(ColumnForm):
    formatting_unfolded = forms.BooleanField(initial=False, required=False)
    formatting_unfolded.widget.attrs.update({"x-model": "open", "class": "hidden"})
    template_name = "columns/forms/column_form.html"

    class Meta:
        model = Column
        fields = (
            "column",
            "part",
            "name",
            "rounding",
            "currency",
            "is_percentage",
            "sort_index",
            "conditional_formatting",
            "positive_threshold",
            "negative_threshold",
        )

        widgets = {
            "currency": Datalist(),
            "sort_index": forms.HiddenInput(),
        }

    ignore_live_update_fields = [
        "currency",
        "name",
        "rounding",
        "positive_threshold",
        "negative_threshold",
        "conditional_formatting",
    ]

    def get_live_fields(self):
        fields = super().get_live_fields()
        fields += ["sort_index"]
        if self.column_type:
            fields += ["name", "formatting_unfolded"]

        if isinstance(self.column_type, (Floating, Integer)):
            fields += [
                "rounding",
                "currency",
                "is_percentage",
                "conditional_formatting",
                "positive_threshold",
                "negative_threshold",
            ]

        return fields


class AggregationColumnForm(SchemaFormMixin, LiveAlpineModelForm):
    class Meta:
        fields = (
            "column",
            "function",
        )
        help_texts = {
            "column": "Select the column to aggregate over",
            "function": "Select the aggregation function",
        }
        model = AggregationColumn
        show = {"function": "column != null"}

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        disable_struct_and_array_columns(
            self.fields, self.fields["column"], self.schema
        )

        agg = {k: [x.value for x in v] for k, v in AGGREGATION_TYPE_MAP.items()}
        self.fields["column"].widget.attrs.update(
            {"x-effect": f"$data.function_choices = {agg}[schema[column]] || []"}
        )


class AggregationFormWithFormatting(AggregationColumnForm):
    formatting_unfolded = forms.BooleanField(initial=False, required=False)
    formatting_unfolded.widget.attrs.update({"x-model": "open", "class": "hidden"})
    template_name = "columns/forms/column_form.html"

    class Meta:
        fields = (
            "column",
            "function",
            "name",
            "rounding",
            "currency",
            "is_percentage",
            "sort_index",
            "conditional_formatting",
            "positive_threshold",
            "negative_threshold",
        )
        help_texts = {
            "column": "Select the column to aggregate over",
            "function": "Select the aggregation function",
        }
        model = AggregationColumn
        widgets = {
            "currency": Datalist(),
            "sort_index": forms.HiddenInput(),
        }

    ignore_live_update_fields = [
        "currency",
        "name",
        "rounding",
        "positive_threshold",
        "negative_threshold",
        "conditional_formatting",
    ]

    def get_live_fields(self):
        fields = super().get_live_fields()
        fields += ["sort_index"]
        if self.column_type:
            if not (
                isinstance(self.parent_instance, Widget)
                and self.parent_instance.kind == Widget.Kind.METRIC
            ):
                fields += ["name"]
            # For aggregation column the numeric type of the output is guaranteed
            fields += [
                "formatting_unfolded",
                "rounding",
                "currency",
                "is_percentage",
                "conditional_formatting",
                "positive_threshold",
                "negative_threshold",
            ]

        return fields


def _get_show_for_function_field(function_type):
    ibis = FUNCTION_TO_IBIS[function_type]
    return f"{ibis}.includes(schema[column])"


def _get_show_for_value_field(value):
    operations = [
        k
        for k, v in AllOperations.items()
        if v.arguments == 1 and v.value_field == value
    ]
    return f"{operations}.includes($data[computed])"


class OperationColumnForm(SchemaFormMixin, LiveAlpineModelForm):
    class Meta:
        model = EditColumn
        fields = (
            "column",
            "string_function",
            "integer_function",
            "date_function",
            "time_function",
            "boolean_function",
            "datetime_function",
            "integer_value",
            "float_value",
            "string_value",
        )

        widgets = {
            "string_value": forms.Textarea(attrs={"rows": 1}),
        }

        show = {
            k: _get_show_for_function_field(k)
            for k in fields
            if k.endswith("_function")
        } | {k: _get_show_for_value_field(k) for k in fields if k.endswith("_value")}

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        disable_struct_and_array_columns(
            self.fields, self.fields["column"], self.schema
        )

        self.fields["column"].widget.attrs.update(
            {"x-effect": f"$data.computed = {IBIS_TO_FUNCTION}[schema[column]]"}
        )

    def save(self, commit: bool):
        # Make sure only one function is set and turn the others to Null
        for field in self.base_fields:
            if field.endswith("function") and f"{self.prefix}-{field}" not in self.data:
                setattr(self.instance, field, None)
        return super().save(commit=commit)


class AddColumnForm(SchemaFormMixin, LiveAlpineModelForm):
    class Meta:
        model = AddColumn
        fields = (
            "column",
            "string_function",
            "integer_function",
            "date_function",
            "time_function",
            "boolean_function",
            "datetime_function",
            "integer_value",
            "float_value",
            "string_value",
            "label",
        )

        widgets = {
            "string_value": forms.Textarea(attrs={"rows": 1}),
        }

        show = (
            {
                k: _get_show_for_function_field(k)
                for k in fields
                if k.endswith("_function")
            }
            | {k: _get_show_for_value_field(k) for k in fields if k.endswith("_value")}
            | {"label": "computed != null"}
        )

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.fields["column"].widget.attrs.update(
            {"x-effect": f"$data.computed = {IBIS_TO_FUNCTION}[schema[column]]"}
        )

    def clean_label(self):
        return column_naming_validation(self.cleaned_data["label"], self.schema.names)


class FormulaColumnForm(BaseLiveSchemaForm):
    class Meta:
        fields = ("formula", "label")
        model = FormulaColumn

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.fields["formula"].widget = CodeMirror(self.schema)

    def clean_label(self):
        return column_naming_validation(self.cleaned_data["label"], self.schema.names)


class WindowColumnForm(BaseLiveSchemaForm):
    class Meta:
        fields = ("column", "function", "group_by", "order_by", "ascending", "label")
        model = WindowColumn

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        disable_struct_and_array_columns(
            self.fields, self.fields["column"], self.schema
        )
        choices = create_column_choices(self.schema)

        if self.column_type is not None:
            self.fields["function"].choices = [
                (choice.value, choice.name)
                for choice in AGGREGATION_TYPE_MAP[self.column_type.name]
            ]
            self.fields["group_by"] = forms.ChoiceField(
                choices=choices,
                help_text=self.base_fields["group_by"].help_text,
                required=False,
                widget=SelectWithDisable(
                    disabled={
                        name: f"You cannot group by a {type_} column"
                        for name, type_ in self.schema.items()
                        if isinstance(type_, Floating)
                    }
                ),
            )
            self.fields["order_by"] = forms.ChoiceField(
                choices=choices,
                help_text=self.base_fields["order_by"].help_text,
                required=False,
            )

    def get_live_fields(self):
        fields = ["column"]

        if self.column_type is not None:
            fields += ["function", "group_by", "order_by", "ascending", "label"]

        return fields

    def clean_label(self):
        return column_naming_validation(self.cleaned_data["label"], self.schema.names)


class ConvertColumnForm(BaseLiveSchemaForm):
    class Meta:
        fields = ("column", "target_type")
        model = ConvertColumn
        labels = {"target_type": "Type"}

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.fields["column"].choices = create_column_choices(self.schema)


def create_left_join_choices(parents, index):
    columns = (
        (
            f"Input {idx+1}",
            sorted(
                [(f"{idx}:{col}", col) for col in parent.schema],
                key=lambda x: str.casefold(x[1]),
            ),
        )
        for idx, parent in enumerate(parents[: index + 1])
    )

    return [("", "No column selected"), *columns]


class JoinColumnForm(BaseLiveSchemaForm):
    class Meta:
        model = JoinColumn
        fields = ["how", "left_column", "right_column"]

    ignore_live_update_fields = ["how", "left_column", "right_column"]

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        parents = kwargs["parent_instance"].parents_ordered.all()
        index = (
            int(index) if (index := self.prefix.split("-")[-1]) != "__prefix__" else 0
        )

        self.fields["left_column"] = forms.ChoiceField(
            choices=create_left_join_choices(parents, index),
            help_text=self.fields["left_column"].help_text.format(
                index + 1 if index == 0 else f"1 to {index+1}"
            ),
        )
        self.fields["right_column"] = forms.ChoiceField(
            choices=create_column_choices(
                parents[index + 1].schema,
            ),
            help_text=self.fields["right_column"].help_text.format(index + 2),
        )

    def get_initial_for_field(self, field, field_name):
        if field_name == "left_column" and self.instance.left_column:
            return f"{self.instance.left_index}:{self.instance.left_column}"
        return super().get_initial_for_field(field, field_name)

    def save(self, *args, **kwargs):
        left_index, left_column = self.cleaned_data["left_column"].split(":")
        self.instance.left_index = int(left_index)
        self.instance.left_column = left_column

        return super().save(*args, **kwargs)


class SortColumnForm(BaseLiveSchemaForm):
    class Meta:
        model = SortColumn
        fields = ["column", "ascending", "sort_index"]
        widgets = {"sort_index": forms.HiddenInput()}


def column_naming_validation(new_name, existing_columns):
    if new_name in (existing_columns):
        raise forms.ValidationError("This column already exists", code="invalid")

    if new_name.lower() in [column.lower() for column in (existing_columns)]:
        raise forms.ValidationError(
            "This column already exists with a different capitalisation", code="invalid"
        )
    return new_name


def extract_values_from_data(prefix, data, field_name, idx=None):
    return tuple(
        value
        for key, value in data.items()
        if key.startswith(prefix)
        and key.endswith(field_name)
        and (idx is None or int(key.split("-")[1]) < idx)
    )


class RenameColumnForm(BaseLiveSchemaForm):
    class Meta:
        model = RenameColumn
        fields = ("column", "new_name")

    def clean_new_name(self):
        formset_prefix, idx = self.prefix.split("-")
        idx = int(idx)
        new_names = extract_values_from_data(formset_prefix, self.data, "new_name", idx)
        old_names = extract_values_from_data(formset_prefix, self.data, "column")

        existing_columns = set(self.schema.names) - set(old_names) | set(new_names)
        return column_naming_validation(
            self.cleaned_data["new_name"],
            list(existing_columns),
        )
