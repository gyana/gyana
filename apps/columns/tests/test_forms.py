import pytest
from django.http import QueryDict

from apps.base.tests.asserts import assertFormChoicesLength
from apps.base.tests.mock_data import TABLE
from apps.base.tests.mocks import mock_bq_client_with_schema
from apps.columns.forms import (
    AddColumnForm,
    AggregationColumnForm,
    AggregationFormWithFormatting,
    ColumnFormWithFormatting,
    ConvertColumnForm,
    FormulaColumnForm,
    JoinColumnForm,
    OperationColumnForm,
    RenameColumnForm,
    WindowColumnForm,
)
from apps.nodes.models import Node

pytestmark = pytest.mark.django_db

COLUMNS_LENGTH = 10


def test_column_form_with_formatting(column_factory, node_factory):
    column = column_factory(node=node_factory())
    form = ColumnFormWithFormatting(instance=column, schema=TABLE.schema())

    assert set(form.fields.keys()) == {"column", "sort_index"}
    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    data = QueryDict(mutable=True)
    data["column"] = "id"
    form = ColumnFormWithFormatting(instance=column, schema=TABLE.schema(), data=data)
    assert set(form.fields.keys()) == {
        "column",
        "sort_index",
        "currency",
        "name",
        "rounding",
        "formatting_unfolded",
        "is_percentage",
        "conditional_formatting",
        "positive_threshold",
        "negative_threshold",
    }

    data["column"] = "athlete"
    form = ColumnFormWithFormatting(instance=column, schema=TABLE.schema(), data=data)
    assert set(form.fields.keys()) == {
        "column",
        "sort_index",
        "name",
        "formatting_unfolded",
    }


def test_aggregation_form(aggregation_column_factory, node_factory, snapshot):
    column = aggregation_column_factory(node=node_factory())
    form = AggregationColumnForm(instance=column, schema=TABLE.schema())

    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    assert {"show": form.show, "effect": form.effect} == snapshot


def test_aggregation_form_with_formatting(aggregation_column_factory, node_factory):
    column = aggregation_column_factory(node=node_factory())
    form = AggregationFormWithFormatting(instance=column, schema=TABLE.schema())

    assert set(form.fields.keys()) == {"column", "sort_index"}
    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    data = QueryDict(mutable=True)
    data["column"] = "id"
    form = AggregationFormWithFormatting(
        instance=column, schema=TABLE.schema(), data=data
    )
    assert set(form.fields.keys()) == {
        "column",
        "sort_index",
        "function",
        "currency",
        "name",
        "rounding",
        "formatting_unfolded",
        "is_percentage",
        "conditional_formatting",
        "positive_threshold",
        "negative_threshold",
    }

    data["column"] = "athlete"
    form = AggregationFormWithFormatting(
        instance=column, schema=TABLE.schema(), data=data
    )
    assert set(form.fields.keys()) == {
        "column",
        "sort_index",
        "function",
        "name",
        "formatting_unfolded",
        "currency",
        "rounding",
        "formatting_unfolded",
        "is_percentage",
        "conditional_formatting",
        "positive_threshold",
        "negative_threshold",
    }


def test_operation_column_form(edit_column_factory, snapshot):
    column = edit_column_factory()
    form = OperationColumnForm(instance=column, schema=TABLE.schema())

    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    assertFormChoicesLength(form, "string_function", 13)
    assertFormChoicesLength(form, "date_function", 6)
    assert {"show": form.show, "effect": form.effect} == snapshot


def test_add_column_form(add_column_factory, snapshot):
    column = add_column_factory()
    form = AddColumnForm(instance=column, schema=TABLE.schema())

    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    assertFormChoicesLength(form, "integer_function", 18)
    assertFormChoicesLength(form, "time_function", 7)
    assert {"show": form.show, "effect": form.effect} == snapshot


def test_formula_form(formula_column_factory):
    column = formula_column_factory()
    form = FormulaColumnForm(instance=column, schema=TABLE.schema())

    assert set(form.fields.keys()) == {"formula", "label"}


def test_window_form(window_column_factory, snapshot):
    column = window_column_factory()
    form = WindowColumnForm(instance=column, schema=TABLE.schema())

    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)
    assertFormChoicesLength(form, "group_by", COLUMNS_LENGTH)
    assertFormChoicesLength(form, "order_by", COLUMNS_LENGTH)

    assert {"show": form.show, "effect": form.effect} == snapshot


def test_convert_form(convert_column_factory):
    column = convert_column_factory()
    form = ConvertColumnForm(instance=column, schema=TABLE.schema())

    assert set(form.fields.keys()) == {"column", "target_type"}
    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)
    assertFormChoicesLength(form, "target_type", 8)


def test_join_form(
    join_column_factory, node_factory, bigquery, integration_table_factory
):
    mock_bq_client_with_schema(
        bigquery, [(name, str(type_)) for name, type_ in TABLE.schema().items()]
    )
    table = integration_table_factory()
    input_node = node_factory(kind=Node.Kind.INPUT, input_table=table)
    second_input = node_factory(kind=Node.Kind.INPUT, input_table=table)
    join_node = node_factory(kind=Node.Kind.JOIN)
    join_node.parents.add(input_node)
    join_node.parents.add(second_input, through_defaults={"position": 1})
    join_node.save()

    column = join_column_factory(node=join_node)
    form = JoinColumnForm(
        instance=column,
        schema=TABLE.schema(),
        parent_instance=column.node,
        prefix="join-column-0",
    )
    assert set(form.fields.keys()) == {
        "left_column",
        "right_column",
        "how",
    }
    assertFormChoicesLength(form, "left_column", 2)
    assertFormChoicesLength(form, "right_column", COLUMNS_LENGTH)
    assertFormChoicesLength(form, "how", 4)


def test_edit_form(edit_column_factory, snapshot):
    column = edit_column_factory()
    form = OperationColumnForm(instance=column, schema=TABLE.schema())

    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    assert {"show": form.show, "effect": form.effect} == snapshot


def test_rename_form(rename_column_factory):
    column = rename_column_factory()
    form = RenameColumnForm(instance=column, schema=TABLE.schema())

    assert set(form.fields.keys()) == {"column", "new_name"}
    assertFormChoicesLength(form, "column", COLUMNS_LENGTH)

    # test validation errors
    prefix = "rename_column-0"
    data = QueryDict(mutable=True)
    data[f"{prefix}-column"] = "lunch"
    data[f"{prefix}-new_name"] = "athlete"

    form = RenameColumnForm(
        instance=column, schema=TABLE.schema(), prefix=prefix, data=data
    )
    form.is_valid()
    assert form.errors["new_name"] == ["This column already exists"]

    data[f"{prefix}-new_name"] = "Athlete"
    form = RenameColumnForm(
        instance=column, schema=TABLE.schema(), prefix=prefix, data=data
    )
    form.is_valid()
    assert form.errors["new_name"] == [
        "This column already exists with a different capitalisation"
    ]

    # Test whether it works with virtual columns
    prefix = "rename_column-1"
    first_prefix = "rename_column-0"
    data = QueryDict(mutable=True)
    data[f"{first_prefix}-column"] = "athlete"
    data[f"{first_prefix}-new_name"] = "brunch"
    data[f"{prefix}-column"] = "lunch"
    data[f"{prefix}-new_name"] = "brunch"

    form = RenameColumnForm(
        instance=column, schema=TABLE.schema(), prefix=prefix, data=data
    )
    form.is_valid()
    assert form.errors["new_name"] == ["This column already exists"]
