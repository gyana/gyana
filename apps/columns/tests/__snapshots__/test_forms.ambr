# serializer version: 1
# name: test_add_column_form
  dict({
    'effect': dict({
      'column': 'computed.function_field = $store.ibis.functions[schema[column]]',
    }),
    'show': dict({
      'boolean_function': "'boolean_function' == $store.ibis.functions[schema[column]]",
      'date_function': "'date_function' == $store.ibis.functions[schema[column]]",
      'datetime_function': "'datetime_function' == $store.ibis.functions[schema[column]]",
      'float_value': "$store.ibis.operations['float_value'].includes($data[computed.function_field])",
      'integer_function': "'integer_function' == $store.ibis.functions[schema[column]]",
      'integer_value': "$store.ibis.operations['integer_value'].includes($data[computed.function_field])",
      'label': 'computed != null',
      'string_function': "'string_function' == $store.ibis.functions[schema[column]]",
      'string_value': "$store.ibis.operations['string_value'].includes($data[computed.function_field])",
      'time_function': "'time_function' == $store.ibis.functions[schema[column]]",
    }),
  })
# ---
# name: test_aggregation_form
  dict({
    'effect': dict({
      'column': 'choices.function = $store.ibis.aggregations[schema[column]]',
    }),
    'show': dict({
      'function': 'column != null',
    }),
  })
# ---
# name: test_edit_form
  dict({
    'effect': dict({
      'column': 'computed.function_field = $store.ibis.functions[schema[column]]',
    }),
    'show': dict({
      'boolean_function': "'boolean_function' == $store.ibis.functions[schema[column]]",
      'date_function': "'date_function' == $store.ibis.functions[schema[column]]",
      'datetime_function': "'datetime_function' == $store.ibis.functions[schema[column]]",
      'float_value': "$store.ibis.operations['float_value'].includes($data[computed.function_field])",
      'integer_function': "'integer_function' == $store.ibis.functions[schema[column]]",
      'integer_value': "$store.ibis.operations['integer_value'].includes($data[computed.function_field])",
      'string_function': "'string_function' == $store.ibis.functions[schema[column]]",
      'string_value': "$store.ibis.operations['string_value'].includes($data[computed.function_field])",
      'time_function': "'time_function' == $store.ibis.functions[schema[column]]",
    }),
  })
# ---
# name: test_operation_column_form
  dict({
    'effect': dict({
      'column': 'computed.function_field = $store.ibis.functions[schema[column]]',
    }),
    'show': dict({
      'boolean_function': "'boolean_function' == $store.ibis.functions[schema[column]]",
      'date_function': "'date_function' == $store.ibis.functions[schema[column]]",
      'datetime_function': "'datetime_function' == $store.ibis.functions[schema[column]]",
      'float_value': "$store.ibis.operations['float_value'].includes($data[computed.function_field])",
      'integer_function': "'integer_function' == $store.ibis.functions[schema[column]]",
      'integer_value': "$store.ibis.operations['integer_value'].includes($data[computed.function_field])",
      'string_function': "'string_function' == $store.ibis.functions[schema[column]]",
      'string_value': "$store.ibis.operations['string_value'].includes($data[computed.function_field])",
      'time_function': "'time_function' == $store.ibis.functions[schema[column]]",
    }),
  })
# ---
# name: test_window_form
  dict({
    'effect': dict({
      'function': '$choices.function = $store.ibis.aggregrations[schema[column]]',
    }),
    'show': dict({
      'ascending': '$data.function != null',
      'function': 'column != null',
      'group_by': '$data.function != null',
      'label': '$data.function != null',
      'order_by': '$data.function != null',
    }),
  })
# ---
